### PR TITLE
add test: check that protected flow allows traffic on main/protected paths

### PR DESCRIPTION
**Pull request by andriidovhan**
ADD tests: 
- Swap flow paths is not causing any packet loss 
- A protected flow allows traffic on main/protected paths

### CODE REVIEW

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy [hunk 1]
```diff
@@ -920,6 +930,86 @@ class ProtectedPathSpec extends BaseSpecification {
...
+            //Minor packet loss is considered a measurement error and happens regardless of reroute
+            assert it.consumerReport.lostPercent < 1
+        }
+
+        and: "Remove the flow"
+        flowHelper.deleteFlow(flow.id)
+    }
+
+    def "A protected flow allows traffic on main/protected paths"() {
+        given: "A flow"
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy]: I think we should specify the type of low. 
`A protected flow`

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy [hunk 2]
```diff
@@ -920,6 +930,86 @@ class ProtectedPathSpec extends BaseSpecification {
...
+        def (Switch srcSwitch, Switch dstSwitch) = topology.activeTraffGens*.switchConnected
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch, true)
+        flow.allocateProtectedPath = true
+        flowHelper.addFlow(flow)
+
+        and: "The flow allows traffic(on a main path)"
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy]: Not native speaker but it looks like it should be `on the main path`.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy [hunk 3]
```diff
@@ -920,6 +930,86 @@ class ProtectedPathSpec extends BaseSpecification {
...
+        def protectedPath = pathHelper.convert(flowPathInfo.protectedPath)
+        northbound.swapFlowPath(flow.id)
+
+        then: "Flow paths are swapped"
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getFlowStatus(flow.id).status == FlowState.UP }
+        def newFlowPathInfo = northbound.getFlowPath(flow.id)
+        pathHelper.convert(newFlowPathInfo) == protectedPath
+        pathHelper.convert(newFlowPathInfo.protectedPath) == mainPath
+
+        and: "The flow allows traffic(on a protected path)"
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy]: Not native speaker but it looks like it should be `on the protected path`.

**Comment by rtretyak** [comment]
Why any of these traffic verifications cannot be added to existing tests?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy [hunk 4]
```diff
@@ -635,11 +643,29 @@ class ProtectedPathSpec extends BaseSpecification {
...
+            Integer numberOfDifferentSwitchInPath = 0
+            for (int i = 0; i < pathSwitches.size() - 1; i++) {
+                if (!pathSwitches[i][1..-2].containsAll(pathSwitches[i + 1][1..-2])) {
+                    numberOfDifferentSwitchInPath += 1
+                }
+            }
+            allLinks.every { link ->
+                !(link.source.switchId == src.dpId && link.destination.switchId == dst.dpId)
+            } && possibleFlowPaths.size() >= 2 && numberOfDifferentSwitchInPath >= 2
+        } ?: assumeTrue("No suiting switches found", false)
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy]: I find the above code close to not maintainable. You are using old ways to obtain switches, which is being deprecated and refactored out. You can achieve the same using new methods in a much more readable and laconic way
```
def tgSwitches = topologyDefinition.getActiveTraffGens()*.getConnectedSwitch()
def (Switch srcSwitch, Switch dstSwitch) = topologyHelper.getAllNotNeighboringSwitchPairs().findAll { 
  def allowsTraffexam = it.src in tgSwitches && it.dst in tgSwitches 
  def usesUniqueSwitches = it.paths.collectMany { pathHelper.getInvolvedSwitches(it) }
                   .unique { it.dpId }.size() > 3
  return allowsTraffexam && usesUniqueSwitches
}
```

Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy]: done

**Comment by sergii-iakovenko** [comment]
@andriidovhan PR has been reverted due to instability of the test. See #2415

**Pull request merged**